Snakefile

```diff
@@ -285,19 +285,31 @@ OPENSUB_LANGUAGE_PAIRS = [
 #
 # Construct this list manually for now to be sure we get the codes in the
 # right order.
-PARALLEL_LANGUAGE_PAIRS = [
+PARALLEL_LANGUAGES = [
     'ar_en', 'cs_en', 'de_en', 'en_es', 'en_fa', 'en_fi', 'en_fr', 'en_id',
     'en_it', 'en_ja', 'en_ko', 'en_lv', 'en_nl', 'en_pl', 'en_pt', 'en_ro',
     'en_ru', 'en_sv', 'en_zh-Hans', 'en_zh-Hant'
 ]
 
 # Tatoeba language pairs; constructed manually such that cs_en, en_lv, en_ro, and
 # en_zh-Hant are removed from PARALLEL_LANGUAGE_PAIRS
-TATOEBA_LANGUAGE_PAIRS = [
+TATOEBA_LANGUAGES = [
     'ar_en', 'de_en', 'en_es', 'en_fa', 'en_fi', 'en_fr', 'en_id', 'en_it',
     'en_ja', 'en_ko', 'en_nl', 'en_pl', 'en_pt','en_ru', 'en_sv', 'en_zh-Hans'
 ]
 
+PARALLEL_LANGUAGE_PAIRS = []
+for pair in PARALLEL_LANGUAGES:
+    lang1, lang2 = pair.split('_')
+    PARALLEL_LANGUAGE_PAIRS.append("{}_{}".format(lang1, lang2))
+    PARALLEL_LANGUAGE_PAIRS.append("{}_{}".format(lang2, lang1))
+
+TATOEBA_LANGUAGE_PAIRS = []
+for pair in TATOEBA_LANGUAGES:
+    lang1, lang2 = pair.split('_')
+    TATOEBA_LANGUAGE_PAIRS.append("{}_{}".format(lang1, lang2))
+    TATOEBA_LANGUAGE_PAIRS.append("{}_{}".format(lang2, lang1))
+
```

Comment:
It looks like these blocks are making sure that all language pairs are represented in both directions. Two questions this raises:

- Does this double the amount of data that exquisite-corpus has to process to build its parallel corpora? For example, does it have to build both `en-fr` and `fr-en` now, and they're the same files with the columns switched?

- The `parallel_sources` function takes in a pair of language codes and finds the appropriate source data, with its language codes in alphabetical order. Is the duplication of data only necessary because these new parallel rules aren't using `parallel_sources` and should be, or is there a new dataflow that `parallel_sources` wouldn't support?

Reply:
On the first question: the amount of data only doubles in the end while producing the joined files. For example, the joined files will have both `en_fr` and `fr_en`. The reason for doing this is to produce alignment files in both directions so that we will have an alignment file for `en_fr.fr` as the source and `en_fr.en` as the target as well.

On the second question: Like mentioned above, data gets doubled only in the end. So, the `parallel_sources` function is still finding appropriate source data with its language codes in alphabetical order.

Reply:
I'm not following. Is the alignment from `en_fr.fr` to `en_fr.en` _different_ from the alignment from `fr_en.fr` to `fr_en.en`?

Reply:
Yes, that's correct. According to fast_align's documentation at https://github.com/clab/fast_align

_fast_align generates asymmetric alignments (i.e., by treating either the left or right language in the parallel corpus as primary language being modeled, slightly different alignments will be generated)._

Reply:
Oh, another reason for producing files in both directions: fast_align produces outputs in the `i-j` _Pharaoh format_, where a pair `i-j` indicates that the i<sup>th</sup> word of the left (source) language is aligned to the j<sup>th</sup> word of the right (target) language. OpenNMT-py expects the alignment file in this format. To feed the right-left alignment to OpenNMT-py, we would have to switch the `i-j` index and generate a new file. Producing alignment files in both directions instead was a safe bet.

Reply:
I'm still losing the line of reasoning here, because I didn't think I was asking about the left versus the right language in my question. `en_fr.fr` contains _exactly the same text_ as `fr_en.fr`, does it not?

Reply:
And I know it's confusing, because `en_fr` and `fr_en` would be similar files but with their columns switched. But `en_fr.fr` and `fr_en.fr` are just the French column of those files, which should be identical.

Reply:
Oh, I get your confusion now. `fr_en.fr` never gets generated. In the `paired` dir, you will have `en_fr.en` and `en_fr.fr` and in the `joined` dir you will have `en_fr` and `fr_en` (similar files but with their columns switched). Hope this makes sense.

Reply:
Okay.

I notice that this changes the global lists of language pairs, including those that are used for other build steps. Do you think the other build steps that use parallel data should just use the duplicated data now, instead of getting their list of inputs through functions that alphabetize the language codes?

Reply:
`PARALLEL_LANGUAGE_PAIRS` is used by `rule parallel` and `rule alignment` to create both directional pairs in `joined` and `alignment` dirs, respectively. Maybe I am missing something here; can you let me know which *global lists of language pairs* is being changed? I will try to make sure that parallel data is not duplicated for those build steps.

Reply:
`PARALLEL_LANGUAGE_PAIRS` is the global I was referring to, but, I recognize that most of the things using it want it to contain both directions now.

I have to say I find the parallel language flow hard to follow (which it already was), so maybe I'm wrong about my specific concern. There probably isn't any problem that this change would introduce. I'm now just wondering if the overall flow could be simplified, given that we're generating alignments in both directions.